File: csharp/luhn/Luhn.cs
public static class Luhn
{
    public static bool IsValid(string number)
    {
        // Spaces should be removed.
        string value = number.Replace(" ", "");

        // Check if too short.
        if (value.Length <= 1)
            return false;

        // depending on if length % 2 == 1 or 0  double on even or odd chars
        // want 2nd from last to be doubled.
        int doubleIt = value.Length % 2;
        int sum = 0;
        for (int i = 0; i < value.Length; i++)
        {
            char ch = value[i];
            // Non-Digit characters are not allowed.
            if (!Char.IsDigit(ch))
                return false;

            // Convert to number.
            int temp = (int)(value[i] - '0');
            // Double if required
            if (i % 2 == doubleIt)
            {
                temp *= 2;
                // Subtract 9 if >= 10
                if (temp >= 10)
                    temp -= 9;
            }
            // Total all the digits.
            sum += temp;
        }
        // True if total is evenly divisible by 10.
        return sum % 10 == 0;
    }
}

File: javascript/luhn/luhn.js
//
// This is only a SKELETON file for the 'Luhn' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const valid = (value) => {
  const digits = '0123456789'
  // check for non-space non-digit characters fail if found.
  let invalidChars = [... value].filter((ch) => ch !== ' ' && !digits.includes(ch))
  if (invalidChars.length > 0) 
    return false

  // convert to an array of numbers removing spaces
  let working = [... value].filter((ch) => digits.includes(ch)).map((ch) => Number(ch))
  // are we checking the odd or even characters
  let flag = working.length % 2

  // if this is too short then fail
  if (working.length <= 1)
    return false

  // map the number doubler, then reduce to a count and return if that count is 
  // evenly divisible by ten.
  return working.map((num, index) => {
    if (index % 2 === flag)
    {
      let temp = num * 2
      if (temp > 9) 
        return temp - 9
      else
        return temp
    }
    return num
  }).reduce((previous, current) => previous + current, 0) % 10 === 0
};
